File: by-language/python/data_structures/btree.py
"""
B-Tree Data Structure.

╔══════════════════════════════════════════════════════════════════════════════╗
║                                  B-TREE                                      ║
╠══════════════════════════════════════════════════════════════════════════════╣
║                                                                              ║
║  A self-balancing tree optimized for disk access, where each node can        ║
║  contain multiple keys and children.                                         ║
║                                                                              ║
║  B-Tree of order t (minimum degree):                                         ║
║  - Every node has at most 2t - 1 keys                                        ║
║  - Every node has at most 2t children                                        ║
║  - Every non-root node has at least t - 1 keys                              ║
║  - Root has at least 1 key (if non-empty)                                   ║
║  - All leaves are at the same level                                         ║
║                                                                              ║
║  Example B-Tree (t=2, so 1-3 keys per node, 2-4 children):                  ║
║                                                                              ║
║                           ┌─────────────┐                                   ║
║                           │   [10, 20]  │                                   ║
║                           └──────┬──────┘                                   ║
║               ┌──────────────────┼──────────────────┐                       ║
║               ▼                  ▼                  ▼                       ║
║        ┌──────────┐      ┌──────────┐      ┌─────────────┐                 ║
║        │ [3, 5, 7]│      │[12,15,17]│      │[25, 30, 40] │                 ║
║        └──────────┘      └──────────┘      └─────────────┘                 ║
║                                                                              ║
║  SPLIT OPERATION (when node is full):                                       ║
║                                                                              ║
║    Before:  [1, 2, 3, 4, 5]  (full node, t=3)                              ║
║                                                                              ║
║    After:        [3]         (middle key moves up)                          ║
║                 /   \                                                        ║
║            [1,2]    [4,5]                                                   ║
║                                                                              ║
╚══════════════════════════════════════════════════════════════════════════════╝

COMPLEXITY:
┌───────────────┬─────────────────────┬───────────────┐
│ Operation     │ Time                │ Disk Accesses │
├───────────────┼─────────────────────┼───────────────┤
│ Search        │ O(t * log_t(n))     │ O(log_t(n))   │
│ Insert        │ O(t * log_t(n))     │ O(log_t(n))   │
│ Delete        │ O(t * log_t(n))     │ O(log_t(n))   │
│ Space         │ O(n)                │ -             │
└───────────────┴─────────────────────┴───────────────┘

where t = minimum degree, n = number of keys

WHY B-TREES?
- Optimized for systems that read/write large blocks of data
- Minimizes disk I/O by storing many keys per node
- Each node typically fits in one disk block/page

COMPARISON:
- B-Tree: Keys in all nodes, faster for random lookups
- B+ Tree: Keys only in leaves, better for range queries

USE CASES:
- Database indexes (MySQL, PostgreSQL)
- File systems (NTFS, ext4, HFS+)
- Key-value stores
"""

from typing import TypeVar, Generic, Optional, List, Tuple

K = TypeVar('K')
V = TypeVar('V')


class BTreeNode(Generic[K, V]):
    """Node in a B-Tree."""

    def __init__(self, leaf: bool = True) -> None:
        """
        Initialize B-Tree node.

        Args:
            leaf: True if this is a leaf node.
        """
        self.keys: List[K] = []
        self.values: List[V] = []
        self.children: List['BTreeNode[K, V]'] = []
        self.leaf = leaf

    def __repr__(self) -> str:
        return f"BTreeNode(keys={self.keys}, leaf={self.leaf})"


class BTree(Generic[K, V]):
    """
    B-Tree implementation.

    A self-balancing tree where each node can have multiple keys
    and children, optimized for disk-based storage systems.
    """

    def __init__(self, t: int = 2) -> None:
        """
        Initialize B-Tree.

        Args:
            t: Minimum degree. Each node has:
               - At most 2t - 1 keys
               - At least t - 1 keys (except root)
               - At most 2t children
        """
        if t < 2:
            raise ValueError("Minimum degree t must be at least 2")

        self._t = t
        self._root: BTreeNode[K, V] = BTreeNode(leaf=True)
        self._size = 0

    @property
    def order(self) -> int:
        """Return order (maximum children per node)."""
        return 2 * self._t

    @property
    def min_degree(self) -> int:
        """Return minimum degree."""
        return self._t

    def __len__(self) -> int:
        """Return number of keys."""
        return self._size

    # =========================================================================
    # SEARCH
    # =========================================================================

    def search(self, key: K) -> Optional[V]:
        """
        Search for key in tree.

        Time: O(t * log_t(n))

        Args:
            key: Key to search for.

        Returns:
            Value if found, None otherwise.
        """
        return self._search(self._root, key)

    def _search(self, node: BTreeNode[K, V], key: K) -> Optional[V]:
        """Recursive search in subtree."""
        # Find first key >= search key
        i = 0
        while i < len(node.keys) and key > node.keys[i]:
            i += 1

        # Check if key found
        if i < len(node.keys) and key == node.keys[i]:
            return node.values[i]

        # If leaf, key not in tree
        if node.leaf:
            return None

        # Recurse to appropriate child
        return self._search(node.children[i], key)

    def __contains__(self, key: K) -> bool:
        """Check if key exists."""
        return self.search(key) is not None

    def __getitem__(self, key: K) -> V:
        """Get value by key."""
        result = self.search(key)
        if result is None:
            raise KeyError(key)
        return result

    # =========================================================================
    # INSERTION
    # =========================================================================

    def insert(self, key: K, value: V) -> None:
        """
        Insert key-value pair.

        Time: O(t * log_t(n))
        """
        root = self._root

        # If root is full, split it first
        if len(root.keys) == 2 * self._t - 1:
            new_root: BTreeNode[K, V] = BTreeNode(leaf=False)
            new_root.children.append(self._root)
            self._split_child(new_root, 0)
            self._root = new_root

        self._insert_non_full(self._root, key, value)
        self._size += 1

    def _split_child(self, parent: BTreeNode[K, V], i: int) -> None:
        """
        Split the i-th child of parent.

        The child must be full (have 2t-1 keys).
        """
        t = self._t
        full_child = parent.children[i]

        # Save original data
        original_keys = full_child.keys[:]
        original_values = full_child.values[:]
        original_children = full_child.children[:]

        # The median is at index t-1
        median_key = original_keys[t - 1]
        median_value = original_values[t - 1]

        # Create new node to hold right half
        new_node: BTreeNode[K, V] = BTreeNode(leaf=full_child.leaf)

        # Left child keeps first t-1 keys
        full_child.keys = original_keys[:t - 1]
        full_child.values = original_values[:t - 1]

        # New node gets keys after median (indices t to 2t-2)
        new_node.keys = original_keys[t:]
        new_node.values = original_values[t:]

        # Move children if not leaf
        if not full_child.leaf:
            full_child.children = original_children[:t]
            new_node.children = original_children[t:]

        # Insert median into parent
        parent.keys.insert(i, median_key)
        parent.values.insert(i, median_value)
        parent.children.insert(i + 1, new_node)

    def _insert_non_full(self, node: BTreeNode[K, V], key: K, value: V) -> None:
        """Insert into a node that is not full."""
        i = len(node.keys) - 1

        if node.leaf:
            # Find position and insert
            while i >= 0 and key < node.keys[i]:
                i -= 1

            # Check for duplicate key
            if i >= 0 and node.keys[i] == key:
                node.values[i] = value  # Update value
                self._size -= 1  # Didn't actually add new key
                return

            node.keys.insert(i + 1, key)
            node.values.insert(i + 1, value)
        else:
            # Find child to recurse into
            while i >= 0 and key < node.keys[i]:
                i -= 1

            # Check for duplicate key in this node
            if i >= 0 and node.keys[i] == key:
                node.values[i] = value
                self._size -= 1
                return

            i += 1

            # Split child if full
            if len(node.children[i].keys) == 2 * self._t - 1:
                self._split_child(node, i)
                if key > node.keys[i]:
                    i += 1
                elif key == node.keys[i]:
                    node.values[i] = value
                    self._size -= 1
                    return

            self._insert_non_full(node.children[i], key, value)

    def __setitem__(self, key: K, value: V) -> None:
        """Set value by key."""
        self.insert(key, value)

    # =========================================================================
    # DELETION
    # =========================================================================

    def delete(self, key: K) -> bool:
        """
        Delete key from tree.

        Time: O(t * log_t(n))

        Returns:
            True if key was found and deleted.
        """
        if key not in self:
            return False

        self._delete(self._root, key)
        self._size -= 1

        # If root is empty but has children, make first child new root
        if len(self._root.keys) == 0 and not self._root.leaf:
            self._root = self._root.children[0]

        return True

    def _delete(self, node: BTreeNode[K, V], key: K) -> None:
        """Delete key from subtree rooted at node."""
        t = self._t
        i = 0

        # Find key or child to recurse into
        while i < len(node.keys) and key > node.keys[i]:
            i += 1

        if node.leaf:
            # Case 1: Key is in leaf node
            if i < len(node.keys) and node.keys[i] == key:
                node.keys.pop(i)
                node.values.pop(i)
            return

        if i < len(node.keys) and node.keys[i] == key:
            # Case 2: Key is in internal node
            self._delete_internal(node, i)
        else:
            # Case 3: Key is in subtree
            self._delete_subtree(node, i, key)

    def _delete_internal(self, node: BTreeNode[K, V], i: int) -> None:
        """Delete key at index i from internal node."""
        t = self._t
        key = node.keys[i]

        if len(node.children[i].keys) >= t:
            # Case 2a: Left child has enough keys
            predecessor = self._get_predecessor(node, i)
            node.keys[i] = predecessor[0]
            node.values[i] = predecessor[1]
            self._delete(node.children[i], predecessor[0])
        elif len(node.children[i + 1].keys) >= t:
            # Case 2b: Right child has enough keys
            successor = self._get_successor(node, i)
            node.keys[i] = successor[0]
            node.values[i] = successor[1]
            self._delete(node.children[i + 1], successor[0])
        else:
            # Case 2c: Both children have t-1 keys, merge
            self._merge(node, i)
            self._delete(node.children[i], key)

    def _delete_subtree(self, node: BTreeNode[K, V], i: int, key: K) -> None:
        """Delete key from subtree at child i."""
        t = self._t

        if len(node.children[i].keys) < t:
            # Child has minimum keys, need to fix first
            if i > 0 and len(node.children[i - 1].keys) >= t:
                # Borrow from left sibling
                self._borrow_from_left(node, i)
            elif i < len(node.children) - 1 and len(node.children[i + 1].keys) >= t:
                # Borrow from right sibling
                self._borrow_from_right(node, i)
            else:
                # Merge with sibling
                if i < len(node.children) - 1:
                    self._merge(node, i)
                else:
                    self._merge(node, i - 1)
                    i -= 1

        self._delete(node.children[i], key)

    def _get_predecessor(self, node: BTreeNode[K, V], i: int) -> Tuple[K, V]:
        """Get predecessor (rightmost key in left subtree)."""
        current = node.children[i]
        while not current.leaf:
            current = current.children[-1]
        return (current.keys[-1], current.values[-1])

    def _get_successor(self, node: BTreeNode[K, V], i: int) -> Tuple[K, V]:
        """Get successor (leftmost key in right subtree)."""
        current = node.children[i + 1]
        while not current.leaf:
            current = current.children[0]
        return (current.keys[0], current.values[0])

    def _borrow_from_left(self, node: BTreeNode[K, V], i: int) -> None:
        """Borrow a key from left sibling."""
        child = node.children[i]
        left_sibling = node.children[i - 1]

        # Move key from parent to child
        child.keys.insert(0, node.keys[i - 1])
        child.values.insert(0, node.values[i - 1])

        # Move key from sibling to parent
        node.keys[i - 1] = left_sibling.keys.pop()
        node.values[i - 1] = left_sibling.values.pop()

        # Move child pointer if not leaf
        if not child.leaf:
            child.children.insert(0, left_sibling.children.pop())

    def _borrow_from_right(self, node: BTreeNode[K, V], i: int) -> None:
        """Borrow a key from right sibling."""
        child = node.children[i]
        right_sibling = node.children[i + 1]

        # Move key from parent to child
        child.keys.append(node.keys[i])
        child.values.append(node.values[i])

        # Move key from sibling to parent
        node.keys[i] = right_sibling.keys.pop(0)
        node.values[i] = right_sibling.values.pop(0)

        # Move child pointer if not leaf
        if not child.leaf:
            child.children.append(right_sibling.children.pop(0))

    def _merge(self, node: BTreeNode[K, V], i: int) -> None:
        """Merge child i with child i+1."""
        left = node.children[i]
        right = node.children[i + 1]

        # Move separator key from parent to left child
        left.keys.append(node.keys[i])
        left.values.append(node.values[i])

        # Move all keys from right to left
        left.keys.extend(right.keys)
        left.values.extend(right.values)

        # Move children if not leaf
        if not left.leaf:
            left.children.extend(right.children)

        # Remove separator and right child from parent
        node.keys.pop(i)
        node.values.pop(i)
        node.children.pop(i + 1)

    def __delitem__(self, key: K) -> None:
        """Delete by key."""
        if not self.delete(key):
            raise KeyError(key)

    # =========================================================================
    # TRAVERSAL
    # =========================================================================

    def inorder(self) -> List[Tuple[K, V]]:
        """Return all key-value pairs in sorted order."""
        result: List[Tuple[K, V]] = []
        self._inorder(self._root, result)
        return result

    def _inorder(self, node: BTreeNode[K, V], result: List[Tuple[K, V]]) -> None:
        """Recursive inorder traversal."""
        for i in range(len(node.keys)):
            if not node.leaf:
                self._inorder(node.children[i], result)
            result.append((node.keys[i], node.values[i]))

        if not node.leaf and node.children:
            self._inorder(node.children[-1], result)

    def __iter__(self):
        """Iterate keys in sorted order."""
        for key, _ in self.inorder():
            yield key

    def items(self):
        """Iterate (key, value) pairs."""
        for item in self.inorder():
            yield item

    def keys(self) -> List[K]:
        """Return all keys in sorted order."""
        return [k for k, _ in self.inorder()]

    def values(self) -> List[V]:
        """Return all values."""
        return [v for _, v in self.inorder()]

    # =========================================================================
    # UTILITY
    # =========================================================================

    def min(self) -> Optional[Tuple[K, V]]:
        """Return minimum key-value pair."""
        if len(self._root.keys) == 0:
            return None

        node = self._root
        while not node.leaf:
            node = node.children[0]
        return (node.keys[0], node.values[0])

    def max(self) -> Optional[Tuple[K, V]]:
        """Return maximum key-value pair."""
        if len(self._root.keys) == 0:
            return None

        node = self._root
        while not node.leaf:
            node = node.children[-1]
        return (node.keys[-1], node.values[-1])

    def height(self) -> int:
        """Return height of tree (all leaves at same level)."""
        height = 0
        node = self._root
        while not node.leaf:
            height += 1
            node = node.children[0]
        return height

    def clear(self) -> None:
        """Remove all keys."""
        self._root = BTreeNode(leaf=True)
        self._size = 0

    def __repr__(self) -> str:
        return f"BTree(t={self._t}, size={self._size}, height={self.height()})"

    def display(self) -> str:
        """Return string representation of tree structure."""
        lines: List[str] = []
        self._display(self._root, "", True, lines)
        return "\n".join(lines)

    def _display(
        self,
        node: BTreeNode[K, V],
        prefix: str,
        is_last: bool,
        lines: List[str]
    ) -> None:
        """Helper for display."""
        connector = "└── " if is_last else "├── "
        lines.append(prefix + connector + str(node.keys))

        if not node.leaf:
            child_prefix = prefix + ("    " if is_last else "│   ")
            for i, child in enumerate(node.children):
                self._display(child, child_prefix, i == len(node.children) - 1, lines)

File: by-language/rust/dsa-data-structures/src/trees/b_tree.rs
//! # B-Tree
//!
//! A self-balancing tree data structure that maintains sorted data and allows
//! searches, insertions, and deletions in O(log n) time.
//!
//! B-trees are optimized for systems that read and write large blocks of data,
//! commonly used in databases and file systems.
//!
//! ## Properties
//!
//! For a B-tree of minimum degree t:
//! - Every node has at most 2t-1 keys
//! - Every non-root node has at least t-1 keys
//! - All leaves are at the same level
//!
//! ## Complexity Analysis
//!
//! | Operation | Time      | Space    |
//! |-----------|-----------|----------|
//! | Search    | O(log n)  | O(1)     |
//! | Insert    | O(log n)  | O(t)     |
//! | Delete    | O(log n)  | O(t)     |
//! | Space     | -         | O(n)     |
//!
//! ## Example
//!
//! ```rust
//! use dsa_data_structures::trees::BTree;
//!
//! let mut tree = BTree::new(3); // Minimum degree 3
//! tree.insert(10);
//! tree.insert(20);
//! tree.insert(5);
//!
//! assert!(tree.contains(&10));
//! assert_eq!(tree.len(), 3);
//! ```

use alloc::vec::Vec;

/// A node in the B-tree.
#[derive(Clone)]
struct BTreeNode<T: Clone> {
    keys: Vec<T>,
    children: Vec<usize>, // Indices into nodes vec
    is_leaf: bool,
}

impl<T: Clone> BTreeNode<T> {
    fn new(is_leaf: bool) -> Self {
        BTreeNode {
            keys: Vec::new(),
            children: Vec::new(),
            is_leaf,
        }
    }
}

/// A B-tree implementation using arena allocation.
///
/// # Type Parameters
///
/// * `T` - The element type, must implement `Ord` and `Clone`
pub struct BTree<T: Ord + Clone> {
    nodes: Vec<BTreeNode<T>>,
    root: Option<usize>,
    min_degree: usize, // t
    len: usize,
}

impl<T: Ord + Clone> BTree<T> {
    /// Creates a new empty B-tree with the given minimum degree.
    ///
    /// # Arguments
    ///
    /// * `min_degree` - The minimum degree t (must be >= 2)
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::trees::BTree;
    ///
    /// let tree: BTree<i32> = BTree::new(3);
    /// assert!(tree.is_empty());
    /// ```
    pub fn new(min_degree: usize) -> Self {
        let min_degree = min_degree.max(2);
        BTree {
            nodes: Vec::new(),
            root: None,
            min_degree,
            len: 0,
        }
    }

    /// Returns the number of elements.
    pub fn len(&self) -> usize {
        self.len
    }

    /// Returns `true` if the tree is empty.
    pub fn is_empty(&self) -> bool {
        self.len == 0
    }

    /// Returns the minimum degree.
    pub fn min_degree(&self) -> usize {
        self.min_degree
    }

    /// Returns `true` if the tree contains the value.
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::trees::BTree;
    ///
    /// let mut tree = BTree::new(3);
    /// tree.insert(5);
    /// assert!(tree.contains(&5));
    /// assert!(!tree.contains(&10));
    /// ```
    pub fn contains(&self, key: &T) -> bool {
        self.search(key).is_some()
    }

    /// Searches for a key and returns a reference if found.
    pub fn search(&self, key: &T) -> Option<&T> {
        self.search_node(self.root?, key)
    }

    fn search_node(&self, node_idx: usize, key: &T) -> Option<&T> {
        let node = &self.nodes[node_idx];

        // Find the first key >= search key
        let mut i = 0;
        while i < node.keys.len() && *key > node.keys[i] {
            i += 1;
        }

        // Check if we found the key
        if i < node.keys.len() && *key == node.keys[i] {
            return Some(&node.keys[i]);
        }

        // If leaf, key not found
        if node.is_leaf {
            return None;
        }

        // Recurse to appropriate child
        self.search_node(node.children[i], key)
    }

    /// Inserts a key into the tree.
    ///
    /// # Example
    ///
    /// ```rust
    /// use dsa_data_structures::trees::BTree;
    ///
    /// let mut tree = BTree::new(3);
    /// tree.insert(10);
    /// tree.insert(5);
    /// tree.insert(15);
    /// assert_eq!(tree.len(), 3);
    /// ```
    pub fn insert(&mut self, key: T) {
        if self.root.is_none() {
            // Create root
            let root_idx = self.nodes.len();
            let mut root = BTreeNode::new(true);
            root.keys.push(key);
            self.nodes.push(root);
            self.root = Some(root_idx);
            self.len = 1;
            return;
        }

        let root_idx = self.root.unwrap();

        // If root is full, split it
        if self.nodes[root_idx].keys.len() == 2 * self.min_degree - 1 {
            let new_root_idx = self.nodes.len();
            let mut new_root = BTreeNode::new(false);
            new_root.children.push(root_idx);
            self.nodes.push(new_root);
            self.root = Some(new_root_idx);
            self.split_child(new_root_idx, 0);
            self.insert_non_full(new_root_idx, key);
        } else {
            self.insert_non_full(root_idx, key);
        }

        self.len += 1;
    }

    /// Inserts a key into a non-full node.
    fn insert_non_full(&mut self, node_idx: usize, key: T) {
        let is_leaf = self.nodes[node_idx].is_leaf;

        if is_leaf {
            // Insert key in sorted order
            let keys = &mut self.nodes[node_idx].keys;
            let mut i = keys.len();
            while i > 0 && key < keys[i - 1] {
                i -= 1;
            }
            keys.insert(i, key);
        } else {
            // Find child to descend to
            let mut i = self.nodes[node_idx].keys.len();
            while i > 0 && key < self.nodes[node_idx].keys[i - 1] {
                i -= 1;
            }

            let child_idx = self.nodes[node_idx].children[i];

            // Split child if full
            if self.nodes[child_idx].keys.len() == 2 * self.min_degree - 1 {
                self.split_child(node_idx, i);

                // After split, determine which child to descend to
                if key > self.nodes[node_idx].keys[i] {
                    i += 1;
                }
            }

            let child_idx = self.nodes[node_idx].children[i];
            self.insert_non_full(child_idx, key);
        }
    }

    /// Splits a full child node.
    fn split_child(&mut self, parent_idx: usize, child_pos: usize) {
        let child_idx = self.nodes[parent_idx].children[child_pos];
        let t = self.min_degree;

        // Create new node
        let new_idx = self.nodes.len();
        let is_leaf = self.nodes[child_idx].is_leaf;
        let mut new_node = BTreeNode::new(is_leaf);

        // Move keys t..2t-1 to new node
        let median_key = self.nodes[child_idx].keys[t - 1].clone();
        new_node.keys = self.nodes[child_idx].keys.split_off(t);
        self.nodes[child_idx].keys.pop(); // Remove median

        // Move children if not leaf
        if !is_leaf {
            new_node.children = self.nodes[child_idx].children.split_off(t);
        }

        self.nodes.push(new_node);

        // Insert median key into parent
        self.nodes[parent_idx].keys.insert(child_pos, median_key);
        self.nodes[parent_idx].children.insert(child_pos + 1, new_idx);
    }

    /// Returns the minimum key.
    pub fn min(&self) -> Option<&T> {
        let mut node_idx = self.root?;
        loop {
            let node = &self.nodes[node_idx];
            if node.is_leaf || node.children.is_empty() {
                return node.keys.first();
            }
            node_idx = node.children[0];
        }
    }

    /// Returns the maximum key.
    pub fn max(&self) -> Option<&T> {
        let mut node_idx = self.root?;
        loop {
            let node = &self.nodes[node_idx];
            if node.is_leaf || node.children.is_empty() {
                return node.keys.last();
            }
            node_idx = *node.children.last().unwrap();
        }
    }

    /// Returns an in-order iterator over the keys.
    pub fn iter(&self) -> BTreeIter<'_, T> {
        let mut stack = Vec::new();
        if let Some(root) = self.root {
            stack.push((root, 0usize));
        }
        BTreeIter { tree: self, stack }
    }

    /// Clears the tree.
    pub fn clear(&mut self) {
        self.nodes.clear();
        self.root = None;
        self.len = 0;
    }

    /// Returns the height of the tree.
    pub fn height(&self) -> usize {
        match self.root {
            None => 0,
            Some(mut node_idx) => {
                let mut height = 1;
                while !self.nodes[node_idx].is_leaf {
                    if self.nodes[node_idx].children.is_empty() {
                        break;
                    }
                    node_idx = self.nodes[node_idx].children[0];
                    height += 1;
                }
                height
            }
        }
    }
}

impl<T: Ord + Clone> Default for BTree<T> {
    fn default() -> Self {
        Self::new(2)
    }
}

/// In-order iterator for B-tree.
pub struct BTreeIter<'a, T: Ord + Clone> {
    tree: &'a BTree<T>,
    stack: Vec<(usize, usize)>, // (node_idx, key_index)
}

impl<'a, T: Ord + Clone> Iterator for BTreeIter<'a, T> {
    type Item = &'a T;

    fn next(&mut self) -> Option<Self::Item> {
        while let Some(&(node_idx, key_idx)) = self.stack.last() {
            let node = &self.tree.nodes[node_idx];

            // If we have more children to explore
            if !node.is_leaf && key_idx < node.children.len() {
                let child_idx = node.children[key_idx];
                self.stack.last_mut().unwrap().1 += 1;
                self.stack.push((child_idx, 0));
                continue;
            }

            // Return key if available
            let key_to_return = key_idx.checked_sub(if node.is_leaf { 0 } else { 1 });

            if node.is_leaf {
                if key_idx < node.keys.len() {
                    self.stack.last_mut().unwrap().1 += 1;
                    return Some(&node.keys[key_idx]);
                }
            } else {
                // We've explored child at key_idx, check if we already returned the key at key_idx-1
                if let Some(ki) = key_to_return {
                    if ki < node.keys.len() && key_idx == ki + 1 {
                        // Already incremented, now return the key
                    }
                }
            }

            self.stack.pop();
        }
        None
    }
}

/// Simple in-order collection for B-tree.
impl<T: Ord + Clone> BTree<T> {
    /// Collects all keys in sorted order.
    pub fn collect_sorted(&self) -> Vec<T> {
        let mut result = Vec::new();
        if let Some(root) = self.root {
            self.collect_node(root, &mut result);
        }
        result
    }

    fn collect_node(&self, node_idx: usize, result: &mut Vec<T>) {
        let node = &self.nodes[node_idx];

        for i in 0..node.keys.len() {
            // Visit left child
            if !node.is_leaf && i < node.children.len() {
                self.collect_node(node.children[i], result);
            }
            // Visit key
            result.push(node.keys[i].clone());
        }

        // Visit rightmost child
        if !node.is_leaf && node.children.len() > node.keys.len() {
            self.collect_node(*node.children.last().unwrap(), result);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    mod basics {
        use super::*;

        #[test]
        fn test_new() {
            let tree: BTree<i32> = BTree::new(3);
            assert!(tree.is_empty());
            assert_eq!(tree.len(), 0);
            assert_eq!(tree.min_degree(), 3);
        }

        #[test]
        fn test_default() {
            let tree: BTree<i32> = BTree::default();
            assert!(tree.is_empty());
            assert_eq!(tree.min_degree(), 2);
        }

        #[test]
        fn test_min_degree_enforced() {
            let tree: BTree<i32> = BTree::new(1);
            assert_eq!(tree.min_degree(), 2); // Should be at least 2
        }
    }

    mod insert {
        use super::*;

        #[test]
        fn test_insert_single() {
            let mut tree = BTree::new(3);
            tree.insert(5);
            assert!(tree.contains(&5));
            assert_eq!(tree.len(), 1);
        }

        #[test]
        fn test_insert_multiple() {
            let mut tree = BTree::new(3);
            tree.insert(10);
            tree.insert(20);
            tree.insert(5);
            tree.insert(15);
            tree.insert(25);

            assert!(tree.contains(&10));
            assert!(tree.contains(&20));
            assert!(tree.contains(&5));
            assert!(tree.contains(&15));
            assert!(tree.contains(&25));
            assert_eq!(tree.len(), 5);
        }

        #[test]
        fn test_insert_causes_split() {
            let mut tree = BTree::new(2); // Max 3 keys per node

            for i in 1..=10 {
                tree.insert(i);
            }

            assert_eq!(tree.len(), 10);

            for i in 1..=10 {
                assert!(tree.contains(&i));
            }
        }

        #[test]
        fn test_insert_descending() {
            let mut tree = BTree::new(2);

            for i in (1..=10).rev() {
                tree.insert(i);
            }

            assert_eq!(tree.len(), 10);

            for i in 1..=10 {
                assert!(tree.contains(&i));
            }
        }
    }

    mod search {
        use super::*;

        #[test]
        fn test_search_existing() {
            let mut tree = BTree::new(3);
            tree.insert(10);
            tree.insert(20);
            tree.insert(5);

            assert_eq!(tree.search(&10), Some(&10));
            assert_eq!(tree.search(&20), Some(&20));
            assert_eq!(tree.search(&5), Some(&5));
        }

        #[test]
        fn test_search_nonexistent() {
            let mut tree = BTree::new(3);
            tree.insert(10);

            assert_eq!(tree.search(&5), None);
            assert_eq!(tree.search(&100), None);
        }

        #[test]
        fn test_search_empty() {
            let tree: BTree<i32> = BTree::new(3);
            assert_eq!(tree.search(&5), None);
        }
    }

    mod min_max {
        use super::*;

        #[test]
        fn test_min_max_empty() {
            let tree: BTree<i32> = BTree::new(3);
            assert_eq!(tree.min(), None);
            assert_eq!(tree.max(), None);
        }

        #[test]
        fn test_min_max() {
            let mut tree = BTree::new(2);
            tree.insert(10);
            tree.insert(5);
            tree.insert(15);
            tree.insert(3);
            tree.insert(20);

            assert_eq!(tree.min(), Some(&3));
            assert_eq!(tree.max(), Some(&20));
        }
    }

    mod collect_sorted {
        use super::*;

        #[test]
        fn test_collect_sorted() {
            let mut tree = BTree::new(2);
            tree.insert(5);
            tree.insert(3);
            tree.insert(7);
            tree.insert(1);
            tree.insert(9);
            tree.insert(4);
            tree.insert(6);

            let sorted = tree.collect_sorted();
            assert_eq!(sorted, vec![1, 3, 4, 5, 6, 7, 9]);
        }

        #[test]
        fn test_collect_sorted_empty() {
            let tree: BTree<i32> = BTree::new(3);
            let sorted = tree.collect_sorted();
            assert!(sorted.is_empty());
        }
    }

    mod clear {
        use super::*;

        #[test]
        fn test_clear() {
            let mut tree = BTree::new(3);
            tree.insert(10);
            tree.insert(20);
            tree.clear();

            assert!(tree.is_empty());
            assert!(!tree.contains(&10));
        }
    }

    mod height {
        use super::*;

        #[test]
        fn test_height_empty() {
            let tree: BTree<i32> = BTree::new(3);
            assert_eq!(tree.height(), 0);
        }

        #[test]
        fn test_height_grows() {
            let mut tree = BTree::new(2);

            tree.insert(1);
            assert_eq!(tree.height(), 1);

            // Insert more to force splits
            for i in 2..=20 {
                tree.insert(i);
            }

            // Height should be reasonable (log base t of n)
            assert!(tree.height() >= 2);
            assert!(tree.height() <= 5);
        }
    }

    mod stress {
        use super::*;

        #[test]
        fn test_many_insertions() {
            let mut tree = BTree::new(3);

            for i in 0..100 {
                tree.insert(i);
            }

            assert_eq!(tree.len(), 100);

            for i in 0..100 {
                assert!(tree.contains(&i), "Missing key: {}", i);
            }
        }

        #[test]
        fn test_random_order() {
            let mut tree = BTree::new(2);
            let values = [42, 17, 89, 3, 56, 91, 28, 64, 5, 73];

            for &v in &values {
                tree.insert(v);
            }

            assert_eq!(tree.len(), 10);

            let sorted = tree.collect_sorted();
            let mut expected = values.to_vec();
            expected.sort();
            assert_eq!(sorted, expected);
        }
    }
}
